File: iptv-player/mobile/lib/domain/logger.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';

import 'redact.dart';

enum LogLevel { debug, info, warn, error }

const _weight = {LogLevel.debug: 10, LogLevel.info: 20, LogLevel.warn: 30, LogLevel.error: 40};

typedef LogWriter = void Function(String line);

void debugLogWriter(String line) => debugPrint(line);

class Logger {
  Logger(this.scope, {this.level = LogLevel.info, this.write = debugLogWriter});

  final String scope;
  final LogLevel level;
  final LogWriter write;

  void debug(String message, [Object? details]) => _write(LogLevel.debug, message, details);
  void info(String message, [Object? details]) => _write(LogLevel.info, message, details);
  void warn(String message, [Object? details]) => _write(LogLevel.warn, message, details);
  void error(String message, [Object? details]) => _write(LogLevel.error, message, details);

  void _write(LogLevel current, String message, Object? details) {
    if (_weight[current]! < _weight[level]!) return;
    final suffix = details == null ? '' : ' ${jsonEncode(redact(details))}';
    write('[${current.name.toUpperCase()}] [$scope] ${redactText(message)}$suffix');
  }
}

File: iptv-player/shared/logger.ts
import { z } from 'zod'
import { redact, redactText } from './redact'
import type { LogLevel } from './types'

const WEIGHT: Record<LogLevel, number> = {
  debug: 10,
  info: 20,
  warn: 30,
  error: 40,
}

export const logLevelSchema = z.enum(['debug', 'info', 'warn', 'error'])

export function resolveLogLevel(options: {
  explicit: string | undefined
  development: boolean
}): LogLevel {
  const parsed = logLevelSchema.safeParse(options.explicit)
  if (parsed.success) return parsed.data
  return options.development ? 'debug' : 'info'
}

export interface Logger {
  debug: (message: string, details?: unknown) => void
  info: (message: string, details?: unknown) => void
  warn: (message: string, details?: unknown) => void
  error: (message: string, details?: unknown) => void
}

function formatDetails(details: unknown, verboseErrors: boolean): string {
  if (details instanceof Error) {
    const text = verboseErrors ? (details.stack ?? details.message) : details.message
    return ` ${redactText(text)}`
  }
  try {
    return ` ${JSON.stringify(redact(details))}`
  } catch {
    return ' [unserializable]'
  }
}

export function createLogger(
  scope: string,
  level: LogLevel,
  verboseErrors = level === 'debug',
): Logger {
  const write = (current: LogLevel, message: string, details?: unknown): void => {
    if (WEIGHT[current] < WEIGHT[level]) return
    const suffix = details === undefined ? '' : formatDetails(details, verboseErrors)
    const line = `[${current.toUpperCase()}] [${scope}] ${redactText(message)}${suffix}`
    if (current === 'error') console.error(line)
    else if (current === 'warn') console.warn(line)
    else console.info(line)
  }

  return {
    debug: (message, details) => write('debug', message, details),
    info: (message, details) => write('info', message, details),
    warn: (message, details) => write('warn', message, details),
    error: (message, details) => write('error', message, details),
  }
}
